### ISSUE TITLE
[Question] Can this crate be used to create a binary program that can duplicate itself with new data included in the new binary?

### ISSUE THREAD

**Issue by benwilber** [gimli-rs/object#547]
This is a little bit of a peculiar use-case, but I'm curious how I would go about doing this.

I want to create a binary program "lua-bundle" that when executed, can create a copy of itself with user-supplied data included in the new binary. I am using mlua to run Lua code with Rust. Instead of running the Lua scripts directly, I actually want the user to be able to "bundle" their Lua code into a new static binary that they can then just run directly. I don't want to require them to use Cargo toolchain (include_str!("...")) etc.

Example usage would be:

$ lua-bundle --include script.lua --output new-bin
$ ./new-bin # Runs the Lua code from script.lua

One idea I had was to declare a static "__PLACEHOLDER__" somewhere in the source lua-bundle binary and then just patch the output binary with the user's included Lua script. But it's really super weird and I don't even know if that would work.  Would work just to append the new data to the end of the binary, but I suspect that will have all kinds of problems with anti-virus, or even just stripping will break it.  I was hoping there was some kind of mechanism with like .data sections or something else I overlooked.

Can this crate be used to do something like that?  If so, is there an example of such a thing?

Thanks!

**Comment by bjorn3** [comment]
Theoretically yes, but you would have to use the low-level writer interfaces (`write::elf::Writer` and `write::pe::Writer`), not the high-level one (`write::Object`) as the latter only support writing object files. I don't think there is an example yet. In any case you did have to do roughly replicate the original executable, except change the data segment to be larger when writing the program table, add the extra data when writing the data segment and make sure that you don't overwrite any other segments when making the data segment larger. Also you can't add data in the middle of a segment. That will mess up all offsets. Instead I did recommend using `enum Void {} extern "C" { static EMBEDDED_DATA: Void; } fn get_embedded_data() -> *const u8 { unsafe { addr_of!(EMBEDDED_DATA).cast() } }`, telling the linker that it is fine to leave `EMBEDDED_DATA` undefined and then when patching the executable fill in the address for `EMBEDDED_DATA` in the dynamic symbol table.

tl;dr: It is complicated and you may need to learn about how the ELF binary format works to implement it.
